tla/ccfraft.tla

```diff
@@ -458,25 +458,25 @@ AppendEntries(i, j) ==
                               log[i][prevLogIndex].term
                           ELSE
                               0
-           \* Send up to 1 entry, constrained by the end of the log.
-           lastEntry == min(Len(log[i]), nextIndex[i][j])
-           entries == SubSeq(log[i], nextIndex[i][j], lastEntry)
-           msg == [mtype          |-> AppendEntriesRequest,
-                   mterm          |-> currentTerm[i],
-                   mprevLogIndex  |-> prevLogIndex,
-                   mprevLogTerm   |-> prevLogTerm,
-                   mentries       |-> entries,
-                   mcommitIndex   |-> min(commitIndex[i], MaxCommittableIndex(SubSeq(log[i],1,lastEntry))),
-                   msource        |-> i,
-                   mdest          |-> j]
-           index == nextIndex[i][j]
+           minIndex == nextIndex[i][j]
+           msg(idx) == 
+               [mtype          |-> AppendEntriesRequest,
+                mterm          |-> currentTerm[i],
+                mprevLogIndex  |-> prevLogIndex,
+                mprevLogTerm   |-> prevLogTerm,
+                mentries       |-> SubSeq(log[i], minIndex, idx),
+                mcommitIndex   |-> min(commitIndex[i], MaxCommittableIndex(SubSeq(log[i],1,idx))),
+                msource        |-> i,
+                mdest          |-> j]
        IN
-       /\ InMessagesLimit(i, j, index)
        /\ messagesSent' =
-            IF Len(messagesSent[i][j]) < index
+            IF Len(messagesSent[i][j]) < minIndex
             THEN [messagesSent EXCEPT ![i][j] = Append(messagesSent[i][j], 1) ]
-            ELSE [messagesSent EXCEPT ![i][j][index] = messagesSent[i][j][index] + 1 ]
-       /\ Send(msg)
+            ELSE [messagesSent EXCEPT ![i][j][minIndex] = messagesSent[i][j][minIndex] + 1 ]
+       \* TODO aft::send_append_entries (raft.h) says is only ever sends entries from a single term!
```

Comment:
See https://github.com/microsoft/CCF/pull/3145 for details. In short, because raft decides whether the accept the AE or not based on its header, the header needs to contain authentic information about the term history it covers. To keep the AE header constant size and compatible, we can only use a single term_of_index and therefore a single term.

Because large amounts of short consecutive terms do not typically happen, or do not coincide with high levels of throughput when they do (an unstable network with lots of elections will not manage many writes), the performance impact is thought to be of little practical concern.

Reply:
Are you saying that the the single term limitation is an implementation detail that the spec can ignore?

Reply:
I was just explaining why the implementation behaved that way. Unless there's a benefit to the spec ignoring that, I would think it's best to keep it as close to implementation behaviour as possible.